onnxruntime/core/providers/webnn/builders/impl/shape_op_builder.cc

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Copyright (c) Intel Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/common.h"
+#include "core/providers/shared/utils/utils.h"
+#include "core/providers/webnn/builders/helper.h"
+#include "core/providers/webnn/builders/model_builder.h"
+#include "core/providers/webnn/builders/op_builder_factory.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace webnn {
+
+class ShapeOpBuilder : public BaseOpBuilder {
+  // Add operator related.
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override ORT_MUST_USE_RESULT;
+
+  // Operator support related.
+ private:
+  bool IsOpSupportedImpl(const InitializedTensorSet& /* initializers */, const Node& node,
+                         const WebnnDeviceType device_type, const logging::Logger& logger) const override;
+};
+
+Status ShapeOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                               const Node& node,
+                                               const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  std::vector<int64_t> input_shape;
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[0], input_shape, logger), "Cannot get shape");
+  const auto rank = static_cast<int32_t>(input_shape.size());
+
+  NodeAttrHelper helper(node);
+  auto true_start = helper.Get("start", 0);
+  auto true_end = helper.Get("end", rank);
+
+  // Deal with negative(s) and clamp.
+  true_start = true_start < 0 ? true_start + rank : true_start;
+  true_start = true_start < 0 ? 0 : ((true_start > rank) ? rank : true_start);
+
+  true_end = true_end < 0 ? true_end + rank : true_end;
+  true_end = true_end < 0 ? 0 : ((true_end > rank) ? rank : true_end);
+
+  emscripten::val new_shape = emscripten::val::array(input_shape);
+
+  // Slice the input shape if start or end attribute exists.
+  new_shape = new_shape.call<emscripten::val>("slice", true_start, true_end);
+
+  emscripten::val desc = emscripten::val::object();
+  desc.set("type", emscripten::val("int64"));
+  emscripten::val dims = emscripten::val::array();
+  auto slice_length = true_end < true_start ? 0 : (true_end - true_start);
+  dims.call<void>("push", slice_length);
+  desc.set("dimensions", dims);
+  emscripten::val output_buffer = emscripten::val::global("BigInt64Array").new_(new_shape);
+  // Since WebNN doesn't support Shape op, we calculate the Shape output and pass values to
+  // WebNN's constant op as workaround.
+  emscripten::val output = model_builder.GetBuilder().call<emscripten::val>("constant", desc, output_buffer);
```

Comment:
@fdwr, if a Shape node is the output of a WebNN graph, WebNN implementation in Chromium will throw: 
`Failed to execute 'buildSync' on 'MLGraphBuilder': The operand with name "y" is not an output operand.`

WebNN implementation doesn't allow a constant as an output operand at https://github.com/fdwr/chromium-src-webnn-dml/blob/dml_sd/third_party/blink/renderer/modules/ml/webnn/ml_graph.cc#L192

Maybe wrapping a slice op would make more sense, any other thoughts?

Reply:
Umm, one limitation of slice is that it doesn't allow 0 size.

Reply:
@fdwr, fyi

Reply:
> Umm, one limitation of slice is that it doesn't allow 0 size.

Are you encountered occurrences of shape where start == end, such that the shape [] would be returned? That shape is a scalar, which should be legal (if so, then this is another occurrence of https://github.com/webmachinelearning/webnn/issues/391).

Reply:
Yeah, that should be another issue, https://github.com/webmachinelearning/webnn/issues/391.

Reply:
I approve of what we have now, but ultimately we will need something more advanced. Seeing the graph below (along with another half dozen models I looked at), Shape is almost always fed into a later Reshape or Expand op (which expect [`MLOperand`](https://www.w3.org/TR/webnn/#mloperand)` reshape(`[`MLOperand`](https://www.w3.org/TR/webnn/#mloperand)` input, `[`sequence`](https://webidl.spec.whatwg.org/#idl-sequence)`<`[`unsigned long`](https://webidl.spec.whatwg.org/#idl-unsigned-long)`?> newShape);`), and that unsigned long should be on the CPU side). So we'll want to keep all those later nodes after the Shape on the CPU provider rather than sending them to WebNN and reading back from the GPU.

![image](https://github.com/microsoft/onnxruntime/assets/1809166/752c41c0-fe4c-4875-835b-2cbdd764bd4d)